test/functional/Valhalla/src_lw5/org/openj9/test/lworld/ValueTypeArrayTests.java

```diff
@@ -0,0 +1,731 @@
+			} else if (srcKind == IFACE_TYPE) {
+				// Meaningless combination
+			} else if (srcKind == VAL_TYPE) {
+				assign((PointV[]) arr, idx, (PointV) src);
+			} else if (srcKind == FLATTENABLE_TYPE) {
+				// Meaningless combination
+			} else {
+				fail("Unexpected source type requested "+srcKind);
+			}
+		} else if (arrKind == FLATTENABLE_TYPE) {
+			if (srcKind == OBJ_TYPE) {
+				// Meaningless combination
+			} else if (srcKind == IFACE_TYPE) {
+				// Meaningless combination
+			} else if (srcKind == VAL_TYPE) {
+				// Meaningless combination
+			} else if (srcKind == FLATTENABLE_TYPE) {
+				assign((PointFV![])arr, idx, (PointFV!) src);
+			} else {
+				fail("Unexpected source type requested "+srcKind);
+			}
+		} else {
+			fail("Unexpected array type requested "+arrKind);
+		}
+	}
+
+	/**
+	 * Represents a null reference.
+	 */
+	static final int NULL_REF = 0;
+
+	/**
+	 * Represents the type <code>Object</code> or <code>Object[]</code>
+	 */
+	static final int OBJ_TYPE = 1;
+
+	/**
+	 * Represents the types <code>SomeIface</code> or <code>SomeIface[]</code>.
+	 * {@link SomeIface} is an interface class defined by this test.
+	 */
+	static final int IFACE_TYPE = 2;
+
+	/**
+	 * Represents the type <code>PointV</code> or <code>PointV[]</code>.
+	 * {@link PointV} is a non-flattenable value type class defined by this test.
+	 */
+	static final int VAL_TYPE = 3;
+
+	/**
+	 * Represents the type <code>PointFV</code> or <code>PointFV[]</code>.
+	 * {@link PointFV} is a flattenable value type class defined by this test.
+	 */
+	static final int FLATTENABLE_TYPE = 4;
+
+	/**
+	 * Convenient constant reference to the <code>ArrayIndexOutOfBoundsException</code> class
+	 */
+	static final Class AIOOBE = ArrayIndexOutOfBoundsException.class;
+
+	/**
+	 * Convenient constant reference to the <code>ArrayStoreException</code> class
+	 */
+	static final Class ASE = ArrayStoreException.class;
+
+	/**
+	 * Convenient constant reference to the <code>NullPointerException</code> class
+	 */
+	static final Class NPE = NullPointerException.class;
+
+	/**
+	 * The expected kind of exception that will be thrown, if any, for an
+	 * assignment to an array whose component type is one of {@link #OBJ_TYPE},
+	 * {@link #IFACE_TYPE}, {@link #VAL_TYPE} or {@link #FLATTENABLE_TYPE} with a source
+	 * of one of those same types or {@link #NULL_REF}.
+	 *
+	 * <p><code>expectedAssignmentExceptions[actualArrayKind][actualSourceKind]</code>
+	 */
+	static Class expectedAssignmentExceptions[][] =
+		new Class[][] {
+						null,									// NULL_REF for array is not a possibility
+						new Class[] {null, null, null, null, null}, // All values can be assigned to Object[]
+						new Class[] {null, ASE,  null, null, null}, // ASE for SomeIface[] = Object
+						new Class[] {null, ASE,  ASE,  null, ASE},  // ASE for PointV[] = PointFV, SomeIface
+						new Class[] {null,  ASE,  ASE,  ASE,  null}, // NPE for PointFV[] = null; ASE for PointFV[] = PointV
```

Comment:
It looks like the first entry in this row changed from `NPE` to `null`.  I think as things stand in the "lw5" world, assigning a null reference to a null-restricted array is expected to result in an `ArrayStoreException`.

Reply:
Upon closer inspection I think I may have just rewritten these tests incorrectly. Something like this results in NPE:
```
NRClass![] c = new NRClass![1];
c[0] = (NRClass)null;
```
Where are you seeing the reference for ArrayStoreException?

Reply:
Earlier versions required a `NullPointerException` to be thrown if a null reference was assigned to an array with a primitive value type (later, null-restricted) component type.  According to the proposal described in the most recent versions of JEP 8316779, an `ArrayStoreException` should be thrown in that case instead.  (I only noticed that while reviewing this pull request.  😢 )

```
Arrays of null-restricted types can be assigned to non-restricted supertypes, and a null check continues to be enforced at run time, similarly to other array storage checks.

Range![] a1 = new Range![3];
a1[0] = new Range(-3, 0);

Range[] a2 = a1;
a2[1] = null; // ArrayStoreException

Object[] a3 = a2;
a3[2] = new Object(); // ArrayStoreException
a3[2] = null; // ArrayStoreException
```

Reply:
I guess if the goal is to have things work the same as lw5 as it is, rather than where the specification seems to be going, the `null` should change to `NPE` in this case.

Reply:
> ```
> NRClass![] c = new NRClass![1];
> c[0] = (NRClass)null;
> ```

Theresa @theresa-m, just to make sure I'm understanding the current behaviour on lw5, what happens for an example like the following?  And does enabling "flattening" affect the behaviour?

```
Object[] c = new NRClass![1];
c[0] = null;
```

Reply:
```
Object[] c = new NRClass![1];
c[0] = null;
```
For this case there is no exception thrown.
As Hang's comment below says there is still work to be done for null restricted arrays.

Reply:
I think that ArrayStoreException can be added quickly though, let me try.

Reply:
I think for now just making the tests pass (having no exception thrown) is the quickest path forward without having full support for nullrestricted arrays. I'll make some notes in the tests about what needs to be done in the future